### ISSUE TITLE
Анализ

### ISSUE THREAD

**Issue by dimacalitca020511-jpg** [dimacalitca020511-jpg/-#1]
"""
Engine Sound Diagnosis — Simple Audio Classifier
------------------------------------------------
Этот скрипт анализирует звук двигателя и определяет,
исправен ли он или имеет неисправность (пример: "normal" / "faulty").

Инструкции:
1. Скачайте несколько WAV-файлов и поместите их в dataset/normal и dataset/faulty.
2. Установите зависимости: pip install -r requirements.txt
3. Запустите обучение:   python engine_sound_diagnosis.py --train
4. Протестируйте файл:  python engine_sound_diagnosis.py --predict example.wav
"""

import os
import argparse
import numpy as np
import librosa
import tensorflow as tf
from tensorflow.keras import layers, models
from sklearn.model_selection import train_test_split
from glob import glob

# === Константы ===
SR = 22050
DURATION = 3.0
N_MELS = 64
DATASET_DIR = "dataset"
MODEL_PATH = "engine_model.h5"

# === Функции ===
def load_audio(path, sr=SR, duration=DURATION):
    y, _ = librosa.load(path, sr=sr, mono=True)
    target_len = int(sr * duration)
    if len(y) > target_len:
        y = y[:target_len]
    else:
        y = np.pad(y, (0, target_len - len(y)))
    return y

def to_melspec(y):
    mel = librosa.feature.melspectrogram(y=y, sr=SR, n_mels=N_MELS)
    mel_db = librosa.power_to_db(mel, ref=np.max)
    mel_db = (mel_db + 80) / 80
    return mel_db

def build_model(input_shape, n_classes):
    m = models.Sequential([
        layers.Conv2D(16, (3,3), activation='relu', input_shape=input_shape),
        layers.MaxPooling2D((2,2)),
        layers.Conv2D(32, (3,3), activation='relu'),
        layers.MaxPooling2D((2,2)),
        layers.Flatten(),
        layers.Dense(64, activation='relu'),
        layers.Dense(n_classes, activation='softmax')
    ])
    m.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return m

def load_dataset():
    files, labels = [], []
    class_names = sorted(os.listdir(DATASET_DIR))
    for idx, cls in enumerate(class_names):
        for f in glob(os.path.join(DATASET_DIR, cls, "*.wav")):
            files.append(f)
            labels.append(idx)
    print(f"[INFO] Найдено {len(files)} файлов, классы: {class_names}")
    return files, np.array(labels), class_names

def prepare_data(files):
    X = []
    for f in files:
        y = load_audio(f)
        mel = to_melspec(y)
        X.append(mel)
    X = np.array(X)[..., np.newaxis]
    return X

# === Основной блок ===
def main(train=False, predict_file=None):
    if train:
        files, labels, class_names = load_dataset()
        X_train, X_test, y_train, y_test = train_test_split(files, labels, test_size=0.2, stratify=labels)

        X_train = prepare_data(X_train)
        X_test = prepare_data(X_test)
        input_shape = X_train.shape[1:]
        model = build_model(input_shape, n_classes=len(class_names))
        model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=15, batch_size=8)
        model.save(MODEL_PATH)
        np.save("class_names.npy", np.array(class_names))
        print("[INFO] Модель сохранена.")
        return

    if predict_file:
        model = tf.keras.models.load_model(MODEL_PATH)
        class_names = np.load("class_names.npy")
        y = load_audio(predict_file)
        mel = to_melspec(y)
        X = np.expand_dims(mel, axis=(0, -1))
        pred = model.predict(X)[0]
        idx = int(np.argmax(pred))
        print("\nРезультат:")
        for i, p in enumerate(pred):
            print(f"{class_names[i]}: {p*100:.2f}%")
        print(f"👉 Предсказание: {class_names[idx]}")
        return

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--train", action="store_true", help="Обучить модель")
    parser.add_argument("--predict", help="Путь к .wav файлу для теста")
    args = parser.parse_args()
    main(train=args.train, predict_file=args.predict)
